### ISSUE TITLE
Windows-client: AD domain join fails

### ISSUE THREAD

**Issue by tsluyter** [DefensiveOrigins/APT-Lab-Terraform#10]
Once the three VMs are up and running and the 1020 seconds wait period has passed for the Windows client system, it tries to join the AD domain "labs.local". This however fails, as per the logs in "C:\WindowsAzure\Logs\Plugins\Microsoft.Compute.JsonADDomainExtension\1.3.6"

```
2020-12-29T12:35:31.6823950Z	[Info]:	Get Domain/Workgroup Information
2020-12-29T12:35:31.6980295Z	[Info]:	Current domain:  (), current workgroup: WORKGROUP, IsDomainJoin: True, Target Domain/Workgroup: labs.local.
2020-12-29T12:35:31.6980295Z	[Info]:	Domain Join Path.
2020-12-29T12:35:31.6980295Z	[Info]:	Current Domain name is empty/null. Try to get Local domain name.
2020-12-29T12:35:31.7136536Z	[Info]:	In AD Domain extension process, the local domain is: ''.
2020-12-29T12:35:31.7136536Z	[Info]:	Domain Join will be performed.
2020-12-29T12:35:31.7449047Z	[Error]:	Try join: domain='labs.local', ou='', user='itadmin@labs.local', option='NetSetupJoinDomain, NetSetupAcctCreate' (#3:User Specified), errCode='1355'.
2020-12-29T12:35:31.7605265Z	[Error]:	Setting error code to 53 while joining domain
2020-12-29T12:35:31.8074438Z	[Error]:	Try join: domain='labs.local', ou='', user='itadmin@labs.local', option='NetSetupJoinDomain' (#1:User Specified without NetSetupAcctCreate), errCode='1355'.
2020-12-29T12:35:31.8074438Z	[Error]:	Setting error code to 53 while joining domain
2020-12-29T12:35:31.8230264Z	[Error]:	Computer failed to join domain 'labs.local' from workgroup 'WORKGROUP'.
2020-12-29T12:35:31.8230264Z	[Info]:	Retrying action after 15 seconds, at attempt 10 out of '10'.
2020-12-29T12:35:46.8392188Z	[Error]:	The AD Domain Join enabling process failed.
2020-12-29T12:35:46.8392188Z	[Error]:	Exception(s) occured while joining Domain 'labs.local'
2020-12-29T12:35:46.8547813Z	[Warn]:	<ReportStatus> [{"version":"1","timestampUTC":"2020-12-29T12:35:46.8392188Z","status":{"name":"ADDomainExtension","operation":"Join Domain/Workgroup","status":"error","code":1,"formattedMessage":{"lang":"en-US","message":"Exception(s) occured while joining Domain 'labs.local'"},"substatus":[{"name":"JoinDomainException for Option 3 meaning 'User Specified'","status":"error","code":1,"formattedMessage":{"lang":"en-US","message":"ERROR - Failed to join domain='labs.local', ou='', user='itadmin@labs.local', option='NetSetupJoinDomain, NetSetupAcctCreate' (#3 meaning 'User Specified'). Error code 1355"}},{"name":"JoinDomainException for Option 1 meaning 'User Specified without NetSetupAcctCreate'","status":"error","code":1,"formattedMessage":{"lang":"en-US","message":"ERROR - Failed to join domain='labs.local', ou='', user='itadmin@labs.local', option='NetSetupJoinDomain' (#1 meaning 'User Specified without NetSetupAcctCreate'). Error code 1355"}}]}}]
2020-12-29T12:35:46.8703720Z	[Info]:	Writing extension status to file: 'C:\Packages\Plugins\Microsoft.Compute.JsonADDomainExtension\1.3.6\Status\0.status', Current Sequence number: '0'
```

Error 1355 / 53 apparently involves the client not being able to resolve the name "labs.local". It's odd, because once the VM is up I can domain join manually just fine.

**Comment by tsluyter** [comment]
Based on some suggestions, I'm wondering whether the host has issues resolving the hostname "labs.local". With any luck we can find a quick way to register that either on the DC in DNS, or with some other filthy trick ;)

For example -> https://www.risual.com/2019/10/azure-domain-join-extension-error-1-meaning-user-specified-without-netsetupacctcreate-error-code-1355/

**Comment by tsluyter** [comment]
Interesting! I've just done a re-deploy of my labs on Azure and I noticed the following, right after deployment.

``` 
PS C:\Users\itadmin> nslookup labs.local
Server:  UnKnown
Address:  10.10.98.10

Name:    labs.local
Address:  10.10.98.10

PS C:\Users\itadmin> resolve-dnsname labs.local
resolve-dnsname : labs.local : DNS name does not exist

PS C:\Users\itadmin> resolve-dnsname -name labs.local
resolve-dnsname : labs.local : DNS name does not exist

PS C:\Users\itadmin> resolve-dnsname -name labs.local -server 10.10.98.10

Name                                           Type   TTL   Section    IPAddress
----                                           ----   ---   -------    ---------
labs.local                                     A      600   Answer     10.10.98.10
```

DNS resolution of the "labs.local" name doesn't work fool-proof without specifying the DC as the server, which apparently affects the automated domain joining. Doing it manually apparently uses another name resolution mech (like the olderfashioned nslookup). 

Then... roughly fifteen minutes later, something changes and resolve-dnsname _can_ resolve the DC / domain name without having to manually forcing with "-server". I'm investigating what the cause of this issue could be.

**Comment by tsluyter** [comment]
I'm doing a quick test where I made two changes:
1. The AD DC is configured to use "1.1.1.1" as DNSForwarder server.
2. The WS01 host is configured to **only** use the AD DC as its DNS server. 

This way Internet connectivity should still work, while also forcing it to only look at the AD DC during the first ten/fifteen minutes of its lifetime.

**Comment by tsluyter** [comment]
Yup that fixed it. I will amend my pull request.

**Comment by tsluyter** [comment]
Addressed in pull request #9 .

**Comment by tsluyter** [comment]
Tackled by completed merge.
